adr/ADR-0003-trustyai-service-deployment-using-operator-pattern.md

```diff
@@ -0,0 +1,245 @@
+
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: trustyai-metrics
+  labels:
+    modelmesh-service: modelmesh-serving
+spec:
+  endpoints:
+    - interval: 4s
+      path: /q/metrics
+      honorLabels: true
+      honorTimestamps: true
+      scrapeTimeout: 3s
+      bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+      targetPort: 8080
+      scheme: http
+      params:
+   	 'match[]':
+ 		 - '{__name__= "trustyai_spd"}'
+ 		 - '{__name__= "trustyai_dir"}'
+      metricRelabelings:
+   	 - action: keep
+ 		 regex: trustyai_.*
+ 		 sourceLabels:
+   		 - __name__
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: trustyai-service
+```
+
+
+The `ServiceMonitor` object targets the TrustyAI Service and specifies how Prometheus should scrape metrics from the service, which includes the path to the metrics endpoint (`/q/metrics`), the interval at which it should scrape the metrics (every 4 seconds), and the type of metrics it should scrape (metrics with names that start with `trustyai_`).
+The selector would also be updated to match the labels of the TrustyAI Service from the Custom Resource.
+The scrape interval and metrics names could potentially also be configurable via the custom resource (with the current values as defaults).
+
+
+A possibility for the service monitor customization is the inclusion of custom values using a nested `ref:` field for `serviceMonitoring`. This would allow users to specify a custom configuration for the `ServiceMonitor` object. For example:
+
+
+```yaml
+apiVersion: trustyai.opendatahub.io/v1
+kind: TrustyAIService
+metadata:
+  name: trustyai-service-example
+  namespace: default
+spec:
+  ...
+  serviceMonitoring:
+    ref:
+      apiVersion: monitoring.coreos.com/v1
+      kind: ServiceMonitor
+      ...
+      spec:
+   	 endpoints:
+ 		 - interval: 15s
+```
+
+If such configuration is not provided, the operator will use the default configuration.
+
+### Route
+
+If deployed on OpenShift, the Operator will also create a `Route` object to expose the TrustyAI Service to external clients. The `Route` object will have the following configuration:
+
+
+```yaml
+kind: Route
+apiVersion: route.openshift.io/v1
+metadata:
+  name: trustyai
+  labels:
+    app: trustyai
+    app.kubernetes.io/name: trustyai-service
+    app.kubernetes.io/part-of: trustyai
+    app.kubernetes.io/version: 0.1.0
+spec:
+  to:
+    kind: Service
+    name: trustyai-service
+  port:
+    targetPort: http
+  tls: null
+```
+
+
+Note that TrustyAI isn't currently implementing HTTPS endpoints, so the `tls` field will be set to `null` for now. Once HTTPS is implemented, the `tls` field will be updated to include the TLS configuration.
+
+### Testing
+
+The testing and CI of the TrustyAI Operator will be performed using the following approaches:
+
+* Unit tests for the Operator code, to ensure that the Operator's functionality is correct.
```

Comment:
i definitely recommend looking at [envtest](https://book.kubebuilder.io/reference/envtest.html) to supplement unit tests. it allows for a nice level of integration style testing to be done virtually.